connectors/services/job_scheduler.py

```diff
@@ -0,0 +1,200 @@
+"""
+import asyncio
+import os
+import time
+
+from connectors.byoc import (
+    SYNC_DISABLED,
+    BYOIndex,
+    ConnectorUpdateError,
+    DataSourceError,
+    JobTriggerMethod,
+    ServiceTypeNotConfiguredError,
+    ServiceTypeNotSupportedError,
+    Status,
+    SyncJobIndex,
+    e2str,
+)
+from connectors.logger import logger
+from connectors.services.base import BaseService
+from connectors.utils import trace_mem
+
+
+class JobSchedulerService(BaseService):
+    def __init__(self, config):
+        super().__init__(config)
+        self.errors = [0, time.time()]
+        self.service_config = self.config["service"]
+        self.trace_mem = self.service_config.get("trace_mem", False)
+        self.idling = self.service_config["idling"]
+        self.hb = self.service_config["heartbeat"]
+        self.connector_index = None
+        self.sync_job_index = None
+        self.trace_mem = self.service_config.get("trace_mem", False)
+
+    def raise_if_spurious(self, exception):
+        errors, first = self.errors
+        errors += 1
+
+        # if we piled up too many errors we raise and quit
+        if errors > self.service_config["max_errors"]:
+            raise exception
+
+        # we re-init every ten minutes
+        if time.time() - first > self.service_config["max_errors_span"]:
+            first = time.time()
+            errors = 0
+
+        self.errors[0] = errors
+        self.errors[1] = first
+
+    def stop(self):
+        self.running = False
+        self._sleeps.cancel()
+        if self.connector_index is not None:
+            self.connector_index.stop_waiting()
+        if self.sync_job_index is not None:
+            self.sync_job_index.stop_waiting()
+
+    async def _process_connector(self, connector):
+        with trace_mem(self.trace_mem):
+            if connector.native:
+                logger.debug(f"Connector {connector.id} natively supported")
+            else:
+                try:
+                    await connector.prepare(self.config)
+                except ServiceTypeNotConfiguredError:
+                    logger.error(
+                        f"Service type is not configured for connector {self.config['connector_id']}"
+                    )
+                    return
+                except ConnectorUpdateError as e:
+                    logger.error(e)
+                    return
+                except ServiceTypeNotSupportedError:
+                    logger.debug(
+                        f"Can't handle source of type {connector.service_type}"
+                    )
+                    return
+                except DataSourceError as e:
+                    await connector.error(e)
+                    logger.critical(e, exc_info=True)
+                    self.raise_if_spurious(e)
+                    return
+
+            try:
+                # the heartbeat is always triggered
+                connector.start_heartbeat(self.hb)
```

Comment:
I think it's sufficient to call heartbeat once here - this whole function will take less than a second to execute normally. Though this can be changed later not to bloat this PR.

Reply:
That will introduce `no. of connectors` writes to `.elastic-connectors` index every `config.idle` seconds.